feat: add web test submodule

helenakallekleiv opened the pull request:
Enable creation of Availiability test through the `web-test` submodule. Availiability test created in sub-module to keep a one-to-one mapping, due to Availiability test being a visible resource in Azure. Ref. [Baseline](https://github.com/equinor/terraform-baseline#one-on-one-mapping)

- The `classic-web-test`- and `standard-web-test`-examples demonstrate the creation of a Classic and Standard Availiability test, respectively. 
- Basic example will not create an Availiability test.

helenakallekleiv commented:
Tests pass 🎉

hknutsen commented:
Will have a look 😄

helenakallekleiv commented:
1. This sounds lika a good idea and a good point, I'll implement the changes.
2. On it.
3. The `count` argument was intended to stop the `azurerm_application_insights_standard_web_test` from being made, when not necessary. I encountered a problem with this when running the test for the `Basic`-example, terraform wanted to create the resource even when it was disabled. Although I think with what you suggested in the first point will make this unnecessary.
4. I'll look into it!

helenakallekleiv commented:
@hknutsen Have a look, all tests pass 🎉

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+variable "geo_locations" {
+  description = "A list of physical locations to run this Web Test."
+  type        = list(string)
+  default     = ["emea-gb-db3-azr"] # North Europe
```
Not sure it's a good idea to have a default value here. I think location is something that should always be set explicitly by the user.

helenakallekleiv replied:
Agreed, should default then be completely omitted or an empty list ( [ ] )?

hknutsen replied:
It depends. I see in docs that it's marked as required, so I guess the best option would be to omit it. Could test if it's possible with an empty list though, but I doubt it.

helenakallekleiv replied:
I'll investigate 👍🏼

helenakallekleiv replied:
I'll omit the default argument for now.

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+    url                              = string
+    body                             = optional(string, null)
+    follow_redirects_enabled         = optional(bool, true)
+    http_verb                        = optional(string, "GET")
```
I see in docs that there is no default value. Have you checked that it's set to "GET" by default if no value is provided. If yes, then I agree that we can set default value to "GET".

hknutsen replied:
This comment was meant for the `http_verb` property only.

helenakallekleiv replied:
The default values set for variable arguments, that are without default values in the docs, were set in the early phase based on insubsea configuration. I just forgot to set them back 😁 But I will check what default value is set when no value is provided 👍🏼

helenakallekleiv replied:
Can confirm that `http_verb` has the default value `"GET"`.

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+
+    content = optional(list(object({
+      content_match      = string
+      ignore_case        = optional(bool, true)
+      pass_if_text_found = optional(bool, false)
```
Any particular reason why these default values are set?

helenakallekleiv replied:
As mentioned above ☝🏼

helenakallekleiv replied:
Default value for both of these are `false`.

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+
+  type = object({
+    expected_status_code        = optional(number, 200)
+    ssl_cert_remaining_lifetime = optional(number, 7)
+    ssl_check_enabled           = optional(bool, true)
```
Any particular reason why these default values are set?

helenakallekleiv replied:
As mentioned above ☝🏼 ☝🏼

helenakallekleiv replied:
I have checked and the default values for `ssl_cert_remaining_lifetime` and `ssl_check_enabled` are `0` and `false`, respectively.

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+variable "retry_enabled" {
+  description = "Is retry on Web Test failure enabled?"
+  type        = bool
+  default     = false
```
Any particular reason why this default value is set?

helenakallekleiv replied:
As mentioned above ☝🏼 ☝🏼 ☝🏼

helenakallekleiv replied:
After investigating, it seems like the default value is `false`. Should we still enable it?

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,115 @@
+variable "enabled" {
+  description = "Is this Web Test enabled?"
+  type        = bool
+  default     = false
```
Should it not be enabled by default? 🤔 Any particular reason why it's not?

helenakallekleiv replied:
Agreed, the reason is the same as the ones above, I'll set it to `true` 😁

hknutsen reviewed (commented):
Tests pass for me as well 🎉 

Did leave a few comments I'd like to get clarified though.

Also: for any dynamic blocks, have you verified whether multiple blocks are supported?

helenakallekleiv commented:
I'll look into if the dynamic blocks are necessary. 

For the variables that do not have given default values in the docs, is it strictly necessary to have default values for them in the config if we want them to be the same as the set default value by Azure?

hknutsen commented:
@helenakallekleiv Regarding the dynamic blocks, the idea is that some blocks are required, some only allow a single block, and some allow any number of blocks. You can refer to the baseline for how to deal with the different types of blocks 🙂

helenakallekleiv commented:
The set default values for variables are based on values listed in Terraform Registry and the default values listed in Azure.

hknutsen commented:
Regarding the `request` block which is required in certain scenarios, consider using a lifecycle argument to ensure its provided when needed, see [example](https://github.com/equinor/terraform-azurerm-databricks/blob/v3.1.1/main.tf#L63-L69).

Have you checked if multiple blocks are supported for any of the block types? For example, is it possible to provide multiple `request` blocks to the standard web test resource?

helenakallekleiv commented:
> Regarding the `request` block which is required in certain scenarios, consider using a lifecycle argument to ensure its provided when needed, see [example](https://github.com/equinor/terraform-azurerm-databricks/blob/v3.1.1/main.tf#L63-L69).

I'll look into it!

> Have you checked if multiple blocks are supported for any of the block types? For example, is it possible to provide multiple `request` blocks to the standard web test resource?

For both `request` and `validation_rules` only one block is supported, but `request` contains a nested `header`-block which could be one or more.

hknutsen commented:
@helenakallekleiv Great findings! 😄 

Here's how I understand it based on your comments, the documentation in the Terraform Registry and by playing around in the Azure Portal:

- A single `request` request block is required for standard web test.
- The `request` block supports one or more **optional** nested `header` blocks.
- A single `validation_rules` block is optional for standard web test.

Agree? 😊

helenakallekleiv commented:
> @helenakallekleiv Great findings! 😄
> 
> Here's how I understand it based on your comments, the documentation in the Terraform Registry and by playing around in the Azure Portal:
> 
>     * A single `request` request block is required for standard web test.
> 
>     * The `request` block supports one or more optional nested `header` blocks.
> 
>     * A single `validation_rules` block is optional for standard web test.
> 
> 
> Agree? 😊

Yes, also `validation_rules` contains a non-repeatable, nested `content`-block 🛩️

helenakallekleiv commented:
@hknutsen Have a look now. In the latest refactor-commit, I added a lifecycle block to the standard test resource, and it works just fine. 

However in the refactor-commit previous to this one, I added a lifecycle block to the classic test resource as well, which returns an error if the request block is present in the classic test module in the example. I added this because I discovered that the classic web test example accepts a `request`-block, even though it's a argument specific to the standard web test.

I removed it from the latest refactor-commit and instead edited the descriptions in the test kind specific variables that they are only supported for a specific test. I believe it's enough with a more specific description, but would like your opinion. 😄

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,122 @@
+}
+
+variable "request" {
+  description = "The request generated by this Web Test. Only supported for Standard Web Test."
```
```suggestion
  description = "The request generated by this Web Test. Required for Standard Web Test."
```

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,122 @@
+    http_verb                        = optional(string, "GET")
+    parse_dependent_requests_enabled = optional(bool, false)
+
+    header = optional(map(object({
```
```suggestion
    header = optional(list(object({
```

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,122 @@
+    header = optional(map(object({
+      name  = string
+      value = string
+    })), {})
```
```suggestion
    })), [])
```

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,122 @@
+    ssl_cert_remaining_lifetime = optional(number, 7)
+    ssl_check_enabled           = optional(bool, true)
+
+    content = optional(list(object({
+      content_match      = string
+      ignore_case        = optional(bool, false)
+      pass_if_text_found = optional(bool, true)
+    })), [])
```
```suggestion
    content = optional(object({
      content_match      = string
      ignore_case        = optional(bool, false)
      pass_if_text_found = optional(bool, true)
    }), null)
```

hknutsen commented on modules/web-test/main.tf:
```diff
@@ -0,0 +1,92 @@
+
+  geo_locations = var.geo_locations
+
+  dynamic "request" {
+    for_each = var.request != null ? [var.request] : []
+
+    content {
+      url                              = request.value["url"]
+      follow_redirects_enabled         = request.value["follow_redirects_enabled"]
+      http_verb                        = request.value["http_verb"]
+      parse_dependent_requests_enabled = request.value["parse_dependent_requests_enabled"]
+
+      dynamic "header" {
+        for_each = request.value["header"]
+
+        content {
+          name  = header.value["name"]
+          value = header.value["value"]
+        }
+      }
+    }
+  }
```
```suggestion
  request {
    url                              = var.request["url"]
    follow_redirects_enabled         = var.request["follow_redirects_enabled"]
    http_verb                        = var.request["http_verb"]
    parse_dependent_requests_enabled = var.request["parse_dependent_requests_enabled"]

    dynamic "header" {
      for_each = var.request["header"]

      content {
        name  = header.value["name"]
        value = header.value["value"]
      }
    }
  }
```

hknutsen commented on modules/web-test/variables.tf:
```diff
@@ -0,0 +1,122 @@
+      name  = string
+      value = string
+    })), {})
+  })
+
+  default = null
```
```suggestion
  })

```

hknutsen reviewed (changes requested):
> We could write the contents of the request and validation_rules variables as individual variables, but then the variable file would be a lot longer. Would like your opinion on this as well 😁

Since `request` is a required block, I think individual variables would make more sense 👍 

Good work, see comments.

helenakallekleiv commented:
Encountered an issue with `standard-web-test` example. On first apply, terraform uses Azure default values for `validation_rules`, and disregarding our set default values. When running manual apply after a few minutes, terraform wants to perform a change, using our default values. 

This happens when no `validation_rules`-block is set in `standard-web-test`-example, and when a `validation_rules`-block is set with the default values.

helenakallekleiv commented:
> Encountered an issue with `standard-web-test` example. On first apply, terraform uses Azure default values for `validation_rules`, and disregarding our set default values. When running manual apply after a few minutes, terraform wants to perform a change, using our default values.
> 
> This happens when no `validation_rules`-block is set in `standard-web-test`-example, and when a `validation_rules`-block is set with the default values.

Solved by changing the argument value for `request_url` from `http` to `https`, in `standard-web-test`-example.

hknutsen reviewed (approved):
Looks good ⭐
